src/query/query_eval.rs

```diff
@@ -31,29 +31,63 @@ fn eval_latest<'a>(
     }
 }
 
-fn eval_lookup<'a>(
+fn eval_test<'a>(
     index: &'a Index,
-    lookup_field: LookupLhs,
-    value: LookupRhs,
+    test: Test,
+    value: Literal,
+    lookup_field: Lookup,
 ) -> Result<Vec<&'a Packet>, QueryError> {
     Ok(index
         .packets
         .iter()
-        .filter(|packet| match lookup_field {
-            LookupLhs::Id => match value {
-                LookupRhs::String(str) => packet.id == str,
-                _ => false
-            },
-            LookupLhs::Name => match value {
-                LookupRhs::String(str) => packet.name == str,
-                _ => false
-            },
-            LookupLhs::Parameter(param_name) => packet.parameter_equals(param_name, &value)
-        })
+        .filter(|packet| lookup_filter(packet, &test, &value,&lookup_field))
         .collect())
 }
 
+fn lookup_filter(packet: &Packet, test: &Test, value: &Literal, lookup: &Lookup) -> bool {
+    match (test, lookup) {
+        (Test::Equal, Lookup::Id) => match value {
+            Literal::String(str) => packet.id == *str,
+            _ => false
+        },
+        (Test::NotEqual, Lookup::Id) => match value {
+            Literal::String(str) => packet.id != *str,
+            _ => false
+        },
+        (_, Lookup::Id) => false,
+        (Test::Equal, Lookup::Name) => match value {
+            Literal::String(str) => packet.name == *str,
+            _ => false
+        },
+        (Test::NotEqual, Lookup::Name) => match value {
+            Literal::String(str) => packet.name != *str,
+            _ => false
+        },
+        (_, Lookup::Name) => false,
+        (test, Lookup::Parameter(param_name)) => packet.test_parameter(param_name, test, value),
+    }
+}
+
 impl Packet {
+    pub fn get_parameter(&self, param_name: &str) ->  Option<Literal> {
+        if let Some(params) = &self.parameters {
+            match params.get(param_name)? {
+                JsonValue::Number(number) => {
+                    Some(Literal::Number(number.as_f64()?))
+                },
+                JsonValue::Bool(bool) => {
+                    Some(Literal::Bool(*bool))
+                },
+                JsonValue::String(string) => {
+                    Some(Literal::String(string))
+                },
+                _ => None // Parameters must be number, bool or string
+            }
+        } else {
+            None
+        }
+    }
```

Comment:
the error/option type propagation in this function feels a bit wonky to me, just want to make sure i understand, if the function fails to get a parameter it will just return a `None` but this silently goes up the chain and just returns `false` up the chain. Should we not give some sort of feedback to the user that the `param_name` doesnt exist?

Equally if it fails to parse a number as a f64 i feel like a lot has gone wrong there for it to silently return no matches to the user

Reply:
On first point, we want this to be silent. If we run a query like `parameter:a == 2` this `get_parameter` function will be run for all packets in the archive which could be hundreds. Some might have that parameter but many won't. What we could do down the line is if this matches none of the packets we could then say something like `param_name` doesn't exist for any packet, or try detect typo in the parameter name or something.

On your second point, I think this should always return a Some. It runs this code https://docs.serde.rs/src/serde_json/number.rs.html#242 where the conversion is from a u64 or i64 to an f64. According to https://doc.rust-lang.org/reference/expressions/operator-expr.html#numeric-cast I think this will always work. If the integer has many digits some of these may be rounded. Seems very unlikely we will bump into this though